agent/src/main/java/kanela/agent/api/instrumentation/InstrumentationBuilder.java

```diff
@@ -140,6 +143,14 @@ public Target onTypesMatching(ElementMatcher<? super TypeDescription> typeMatche
         return target;
     }
 
+    public void when(final ClassRefiner.Builder refiner, final Runnable thunk) {
+        val classLoaderMatcher = RefinedClassLoaderMatcher.from(Option.of(ClassLoaderRefiner.from(refiner.build())));
+        if(classLoaderMatcher.matches(ClassLoader.getSystemClassLoader())) {
```

Comment:
This could be a problem because in certain situations (like when running from SBT or with Play applications) the actual user libraries are not part of the system ClassLoader and this will always end up assuming that certain instrumentation should not be loaded. 

Would it be possible to somehow "save" the refiner definition and then apply it later on when the transformers are actually evaluated? Maybe by creating a new ClassLoader matcher that uses the refiners and hook it up [here](https://github.com/kamon-io/kanela/blob/master/agent/src/main/java/kanela/agent/builder/KanelaAgentBuilder.java#L117-L130)?

Reply:
I understand the point, but maybe this approach it is not enough. At first all the instrumentation was lazy and the end of everything the target class was analyzed and generated in matcher classloader passing it here: https://github.com/kamon-io/kanela/blob/master/agent/src/main/java/kanela/agent/builder/KanelaAgentBuilder.java#L70    I am thinking of an alternative...